transport/stunnel/client.go

```diff
@@ -0,0 +1,261 @@
+package stunnel
+
+import (
+	"bytes"
+	"context"
+	"strconv"
+	"text/template"
+
+	"github.com/backube/pvc-transfer/transport"
+	"github.com/backube/pvc-transfer/transport/tls/certs"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const clientListenPort = 6443
+
+const (
+	stunnelClientConfTemplate = `
+ pid =
+ sslVersion = TLSv1.3
+ client = yes
+ syslog = no
+ output = /dev/stdout
+ [rsync]
+ debug = 7
+ accept = {{ .listenPort }}
+ cert = /etc/stunnel/certs/tls.crt
+ key = /etc/stunnel/certs/tls.key
+ CAfile = /etc/stunnel/certs/ca.crt
+ verify = 2
+{{- if not (eq .proxyHost "") }}
+ protocol = connect
+ connect = {{ .proxyHost }}
+ protocolHost = {{ .hostname }}:{{ .listenPort }}
+{{- if not (eq .proxyUsername "") }}
+ protocolUsername = {{ .proxyUsername }}
+{{- end }}
+{{- if not (eq .proxyPassword "") }}
+ protocolPassword = {{ .proxyPassword }}
+{{- end }}
+{{- else }}
+ connect = {{ .hostname }}:{{ .connectPort }}
+{{- end }}
+`
+)
+
+type client struct {
+	logger         logr.Logger
+	ingressPort    int32
```

Comment:
I'm struggling a bit w/ the `ingressPort` name. It's `ConnectPort` in the Transport interface, but the `NewClient` params also refer to it as ingressPort. And Endpoint has IngressPort().
Did we make a mistake by having Transport.ConnectPort() and Endpoint.IngressPort()?

Reply:
It's part of a bad refactor, it should be connectPort here in the transport package.

IMO, having ListenPort and ConnectPort at the transport level makes sense because out of the port one is used to accept connection other is used to connect.

In the case of endpoint interface both are used to accept connections, the difference is whether it is exposed external to the cluster or it is internal because of which BackendPort and IngressPort made much more sense, assuming Ingress is the terms used widely to accept traffic into the cluster. Most Kube resources use targetPort and port to mean the same thing respectively but that was turning out to be a little confusing when we were having conversations in the team about this. I would be open to renaming the endpoint interface methods if we want to be consistent with Kube resource fields